core/src/main/java/org/elasticsearch/transport/CompressibleBytesOutputStream.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.common.compress.CompressorFactory;
+import org.elasticsearch.common.io.Streams;
+import org.elasticsearch.common.io.stream.BytesStream;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.lease.Releasable;
+
+import java.io.IOException;
+
+public class CompressibleBytesOutputStream extends StreamOutput implements Releasable {
+
+    private final StreamOutput stream;
+    private final BytesStream bytesStreamOutput;
+    private final boolean shouldCompress;
+
+    public CompressibleBytesOutputStream(BytesStream bytesStreamOutput, boolean shouldCompress) throws IOException {
+        this.bytesStreamOutput = bytesStreamOutput;
+        this.shouldCompress = shouldCompress;
+        if (shouldCompress) {
+            this.stream = CompressorFactory.COMPRESSOR.streamOutput(Streams.flushOnCloseStream(bytesStreamOutput));
+        } else {
+            stream = bytesStreamOutput;
+        }
+    }
+
+    /**
+     * This method ensures that compression is complete and returns the underlying bytes.
+     *
+     * @return bytes underlying the stream
+     * @throws IOException if an exception occurs when writing or flushing
+     */
+    public BytesStream finishStream() throws IOException {
```

Comment:
What about:
```
public BytesReference materialize() throws IOException
```

I just kind of feel weird about `materializeAndClose` because calling that does not release you from the obligation that you still must call the regular `close()`.